### PR TITLE
EZP-24960: Create the YUI combo controller

### PR DESCRIPTION
**Pull request by michalpipa**
> JIRA issue: https://jira.ez.no/browse/EZP-24960
> Status: ready for review

Controller for combo loader. Combines JS ans CSS files.

Needs to be merged after #413.

### CODE REVIEW

File: Controller/PlatformUIController.php [hunk 1]
```diff
@@ -33,6 +42,28 @@ public function shellAction()
...
     }
 
+    /**
+     * Load JS or CSS assets.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function yuiCombineLoaderAction(Request $request)
```
Comment (bdunogier) [inline Controller/PlatformUIController.php]: I think I would have created a new controller.

File: Resources/config/default_settings.yml [hunk 2]
```diff
@@ -1,6 +1,8 @@
 parameters:
     ez_platformui.default.yui: {}
     ez_platformui.default.yui.filter: "min"
+    ez_platformui.default.yui.combine: true
+    ez_platformui.default.yui.build_dir: /bundles/ezplatformuiassets/vendors/yui3/build/
```
Comment (bdunogier) [inline Resources/config/default_settings.yml]: Hmmm, odd question maybe, but why do we need a siteaccess aware setting here ? PlatformUI itself isn't really siteaccess aware...

File: Loader/ComboLoader.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+ * File containing the ComboLoader class.
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\PlatformUIBundle\Loader;
+
+use eZ\Publish\Core\Base\Exceptions\NotFoundException;
+
+class ComboLoader implements Loader
```
Comment (bdunogier) [inline Loader/ComboLoader.php]: I'd like a bit of phpdoc about this implementation of the Loader interface, and what it does in particular.

File: Controller/PlatformUIController.php (hunk 1 above)
Comment (michalpipa) [inline Controller/PlatformUIController.php]: I was thinking about it. I decided to leave it here, because shell action has similar purpose (admin UI loading).

But no problem for me, I can move this to separate file.

File: Resources/config/default_settings.yml (hunk 2 above)
Comment (dpobel) [inline Resources/config/default_settings.yml]: well, the thing is the whole JS module settings system was made siteaccess aware so for consistency I guess that's OK

File: Loader/Loader.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * JS and CSS loader.
+ */
+interface Loader
+{
+    /**
+     * @param array $files Array of file names to be loaded.
+     *
+     * @return array Array which elements are JS/CSS content and content type
```
Comment (bdunogier) [inline Loader/Loader.php]: I'd really rather have two functions, one to get the content, and another one to get the type.

I know that the type is detected while iterating over the items, but you could very easily extract this part, and only check the first (or last) one.

**Comment by dpobel** [comment]
I just tested and it works well :) even though it breaks some asynchronous loading done by AlloyEditor/CKEditor. I made a quick and dirty workaround in https://github.com/ezsystems/PlatformUIBundle/commit/658598b81a2bd9f09c5ea095884e678fd4458f96 but while you are fixing the issues mentioned by @bdunogier and working on the last sub-task of the story I'll try to find a cleaner solution to that.

File: Controller/PlatformUIController.php (hunk 1 above)
Comment (bdunogier) [inline Controller/PlatformUIController.php]: Don't bother, we can do that later.

File: Resources/config/default_settings.yml (hunk 2 above)
Comment (bdunogier) [inline Resources/config/default_settings.yml]: Yeah... we've talked about it, and it's indeed better to favour consistency here. But the fact is that overall, I'm expecting bad surprises if somebody relies on the siteaccess awareness of PlatformUI :-)

**Comment by bdunogier** [comment]
Looks good, with the remarks taken into account.

Could you please add at least a unit test for the loader method(s) ? Might be even better to cover this with a BDD scenario, but at least a unit test.

**Comment by michalpipa** [comment]
@dpobel Do we need to check if any file was passed to this controller? If yes, then what should it return in case of empty files array? HTTP 400 status code?

We can also assume that there always will be at least one file passed.

**Comment by dpobel** [comment]
I guess you should check if at least a file is passed and indeed a 400 if no file is provided seems to be the correct status code.

**Comment by michalpipa** [comment]
Rebased because of conflicts. Also added unit test and check if array of files to combine is not empty.

File: Loader/Loader.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * Returns combined JS or CSS files.
+     *
+     * @param array $files File name array
+     *
+     * @return string Combined files
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    public function getContent(array $files);
```
Comment (andrerom) [inline Loader/Loader.php]: Maybe `getFileContent`

File: Loader/Loader.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Returns content type.
+     *
+     * @param array $files File name array
+     *
+     * @return string
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    public function getContentType(array $files);
```
Comment (andrerom) [inline Loader/Loader.php]: Like above, mainly to avoid using the domain names from repository, something like `getCombinedFileContentType`, or `getMediaType` or ..

**Comment by andrerom** [comment]
@dpobel  is this supposed to support \@import ?

**Comment by dpobel** [comment]
@andrerom not really but actually we won't use the combo loader for CSS but YUI dynamically load some CSS with it that's why we need to have a minimal support for CSS.

**Comment by andrerom** [comment]
> @andrerom not really but actually we won't use the combo loader for CSS but YUI dynamically load some CSS with it that's why we need to have a minimal support for CSS.

ok, let's make sure to document that then so we don't end up getting feature requests for something we don't intend to use it for *(there are anyway better solutions for this out there)*.

File: Controller/PlatformUIController.php [hunk 7]
```diff
@@ -8,16 +8,27 @@
...
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
 
 class PlatformUIController extends Controller
 {
     /** @var Provider */
     private $configAggregator;
 
-    public function __construct(Provider $configAggregator)
+    /** @var  Loader */
```
Comment (lolautruche) [inline Controller/PlatformUIController.php]: CS: extra WS

File: Controller/PlatformUIController.php (hunk 1 above)
Comment (lolautruche) [inline Controller/PlatformUIController.php]: I wouldn't prefix with `yui` as there is nothing specific to YUI, at least code wise.

File: Loader/ComboLoader.php [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        $this->yuiBuildDirectory = $yuiBuildDirectory;
+        $this->webDirectory = $webDirectory;
+    }
+
+    /**
+     * Returns combined JS or CSS files.
+     *
+     * @param array $files File name array
+     *
+     * @return string Combined files
```
Comment (andrerom) [inline Loader/ComboLoader.php]: "The content of the combined files"

File: Loader/Loader.php (hunk 5 above)
Comment (lolautruche) [inline Loader/Loader.php]: +1

**Comment by andrerom** [comment]
Besides open nitcpicks; +1

File: Controller/PlatformUIController.php (hunk 7 above)
Comment (bdunogier) [inline Controller/PlatformUIController.php]: CS as well: please use the FQN in PhpDoc (phpstorm tip: ctrl+shift+alt+c on the class name, then paste).

Comment (michalpipa) [inline Controller/PlatformUIController.php]: Great tip! Thanks.

File: Loader/ComboLoader.php [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+                $content .= $this->loadFile($file, $ext);
+            } elseif ($ext === 'css') {
+                $file = $this->fixUnresolvedPath($file);
+
+                $content .= $this->loadFile($file, $ext);
+            } else {
+                throw new NotFoundException('file type', $file);
+            }
+        }
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: A `switch` would have been more readable IMHO.

File: Loader/ComboLoader.php [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+     * @param $file
+     * @param $ext
+     *
+     * @return string
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    private function loadFile($file, $ext)
+    {
+        $filename = $this->fixFilename($file, $ext);
+        $fileContent = @file_get_contents($filename);
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: Could you please use Filesystem component instead?

File: Loader/ComboLoader.php [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+        $ext = $this->findFileExtension($file);
+
+        if ($ext === 'js') {
+            $type = 'application/javascript';
+        } elseif ($ext === 'css') {
+            $type = 'text/css';
+        } else {
+            throw new NotFoundException('file type', $file);
+        }
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: Switch as well maybe

File: Loader/ComboLoader.php [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+     *
+     * @return string
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    public function getContentType(array $files)
+    {
+        $this->assertAtLeastOneFile($files);
+
+        $file = key($files);
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: Why `key()`? This will return the index of the current element. Shouldn't you use `reset()` or `current()` instead?

File: Loader/ComboLoader.php [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+            if ($ext === 'js') {
+                if ($this->isYuiFile($file)) {
+                    $file = $this->fixUnresolvedPath($file);
+                }
+
+                $content .= $this->loadFile($file, $ext);
+            } elseif ($ext === 'css') {
+                $file = $this->fixUnresolvedPath($file);
+
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: empty line not needed.

File: Loader/ComboLoader.php [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+     * Replace underscores by dots.
+     *
+     * @param $filename
+     * @param $ext
+     *
+     * @return string
+     */
+    private function fixFilename($filename, $ext)
+    {
+        return $this->webDirectory . str_replace('_' . $ext, '.' . $ext, $filename);
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: By making `yuiBuildDirectory` relative to `webDirectory`, you wouldn't have to add the webdirectory here :wink:

File: Loader/ComboLoader.php (hunk 12 above)
Comment (michalpipa) [inline Loader/ComboLoader.php]: Well, this is because `files` array has file names stored in keys. But I could be a nice idea to swap keys and values in the controller and pass them to loader in values.

Comment (lolautruche) [inline Loader/ComboLoader.php]: Yes, it would make this code more obvious.

File: Loader/ComboLoader.php [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    private function loadFile($file, $ext)
+    {
+        $filename = $this->fixFilename($file, $ext);
+        $fileContent = @file_get_contents($filename);
+
+        if ($fileContent === false) {
+            throw new NotFoundException('file', $file);
+        }
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: Instead of using `@` silent operator, first check if file exists. If not, throw the exception.

File: Loader/ComboLoader.php (hunk 9 above)
Comment (michalpipa) [inline Loader/ComboLoader.php]: I think that for two choices `if/elseif/else` is still readable, but if you insist I can change it.

Comment (lolautruche) [inline Loader/ComboLoader.php]: I'm not saying it's not readable at all, just that it *would* be *more* readable with a switch.
But this is not critical at all...

File: Loader/ComboLoader.php (hunk 14 above)
Comment (michalpipa) [inline Loader/ComboLoader.php]: But `yuiBuildDirectory` is relative to `webDirectory` (it's used only when path is not fully resolved). Can you explain more what do you mean?

Comment (lolautruche) [inline Loader/ComboLoader.php]: I mean that in the service definition, you could pass only one parameter. But currently the YUI build directory is a siteaccess aware setting, so it's not possible as is. 
So then, questions are:
* Does this parameter need to be siteaccess aware?
* If not (which is what I think), why not building a single parameter like this: `ez_platformui.default.yui.build_dir: %webroot_dir%/bundles/ezplatformuiassets/vendors/yui3/build/`? This should simplify the code a bit.

Comment (michalpipa) [inline Loader/ComboLoader.php]: @lolautruche `yuiBuildDirectory` is not always used. 

YUI file path is not fully resolved, so we need to prefix it with `webDirectory` and `yuiBuildDirectory`. But custom file path is fully resolved and it needs `webDirectory` prefix only.

Ref: https://github.com/michalpipa/PlatformUIBundle/blob/f4f9c1ab350b48ef0071635f4c4a60ee9f02872d/Loader/ComboLoader.php#L48-L52

Comment (lolautruche) [inline Loader/ComboLoader.php]: ok then

File: Loader/ComboLoader.php [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+     *
+     * @return string
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    private function loadFile($file, $ext)
+    {
+        $filename = $this->fixFilename($file, $ext);
+
+        if (is_readable($filename)) {
+            $fileContent = @file_get_contents($filename);
```
Comment (lolautruche) [inline Loader/ComboLoader.php]: Please remove the `@` silent operator. It's useless here.

Comment (michalpipa) [inline Loader/ComboLoader.php]: Thanks, I forgot to remove it.

File: Loader/ComboLoader.php (hunk 14 above)
Comment (andrerom) [inline Loader/ComboLoader.php]: But his argument that `ez_platformui.default.yui.build_dir` does not need to be siteaccess aware is still valid.
either you override the yui folder arg or you don't, no value in doing it per siteaccess.

Comment (michalpipa) [inline Loader/ComboLoader.php]: @andrerom We've discussed this: https://github.com/ezsystems/PlatformUIBundle/pull/427#discussion_r44634948

But if needed, I can change this.

File: Resources/config/default_settings.yml (hunk 2 above)
Comment (lolautruche) [inline Resources/config/default_settings.yml]: > well, the thing is the whole JS module settings system was made siteaccess aware so for consistency I guess that's OK

Yes, but here it is completely useless

File: Controller/PlatformUIController.php [hunk 17]
```diff
@@ -33,6 +44,33 @@ public function shellAction()
...
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function combineLoaderAction(Request $request)
+    {
+        $files = array_keys($request->query->all());
+
+        try {
+            $content = $this->loader->getCombinedFileContent($files);
+            $type = $this->loader->getCombinedFileContentType($files);
```
Comment (bdunogier) [inline Controller/PlatformUIController.php]: I'd actually get the types first, and the contents after that: if there is an error in the types,

File: Loader/ComboLoader.php [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+    public function getCombinedFileContent(array $files)
+    {
+        $this->assertAtLeastOneFile($files);
+
+        $content = '';
+
+        foreach ($files as $file) {
+            $ext = $this->findFileExtension($file);
+
+            if ($ext === 'js') {
```
Comment (bdunogier) [inline Loader/ComboLoader.php]: As a sequel to the comment above about getting the type first, then the content.

We could maybe simplify this a bit. Do the type consistency check in `getCombinedFilesContentType` only, and assume in this `getCombinedFileContent` that the list is consistent. We can update this method's description and say that it combines a valid list of files.

This would restrict type consistency checks in one place, and this method would become much simpler.

Comment (crevillo) [inline Loader/ComboLoader.php]: i'd add we could change 'js' and 'css' to constants?

File: Loader/ComboLoader.php [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+     *
+     * @return string
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    private function loadFile($file, $ext)
+    {
+        $filename = $this->fixFilename($file, $ext);
+
+        if (is_readable($filename)) {
+            $fileContent = file_get_contents($filename);
```
Comment (crevillo) [inline Loader/ComboLoader.php]: is the else needed? i mean, can't we ```return file_get_contents($filename)``` and throw without the else?

Comment (bdunogier) [inline Loader/ComboLoader.php]: Good remark. Return asap works well in that case.

File: Loader/Loader.php [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+     * Returns combined JS or CSS files.
+     *
+     * @param array $files File name array
+     *
+     * @return string The content of the combined files
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\Core\Base\Exceptions\NotFoundException
+     */
+    public function getCombinedFileContent(array $files);
```
Comment (bdunogier) [inline Loader/Loader.php]: Naming: instead of "getCombined", we could go for the "combine" verb: ``combineFilesContent` (there are multiple files).

File: Resources/config/routing.yml [hunk 21]
```diff
@@ -14,3 +14,8 @@ template_yui_module:
 ez_pjax:
     resource: "@eZPlatformUIBundle/Resources/config/routing_pjax.yml"
     prefix:   /pjax
+
+yui_combo_loader:
+    path: /combo
```
Comment (bdunogier) [inline Resources/config/routing.yml]: Should we go for something a bit less likely to collide ? If a content is named "Combo", we're gonna have issues :-)

File: Loader/ComboLoader.php (hunk 9 above)
Comment (bdunogier) [inline Loader/ComboLoader.php]: Or we could remove type check from this method, and get rid of the control structure altogether :-)

File: Resources/config/routing.yml (hunk 21 above)
Comment (michalpipa) [inline Resources/config/routing.yml]: That is true. I've used route name suggested in Jira issue.

Any suggestions? /cc @dpobel @andrerom

Comment (lolautruche) [inline Resources/config/routing.yml]: `/_ez_combo_resources`?

Comment (bdunogier) [inline Resources/config/routing.yml]: Why not.

Comment (dpobel) [inline Resources/config/routing.yml]: from a performance perspective, it's better if we keep it short, `/_combo` or `/_ezcombo` or whatever :)

**Comment by bdunogier** [comment]
What are the open issues left on this ?

**Comment by michalpipa** [comment]
> What are the open issues left on this ?

No issues left.

**Comment by lolautruche** [comment]
+1

**Comment by andrerom** [comment]
Failing:
>  [eZ Rest Model tests]: Passed: 17 Failed: 1 Total: 18 (ignored 0) (0.223 seconds)

Any insight on this?

**Comment by dpobel** [comment]
weird failure (and weird report), this has nothing to do with this PR, I'm restarting the job.

**Comment by dpobel** [comment]
@michalpipa: can you cherry-pick https://github.com/ezsystems/PlatformUIBundle/commit/658598b81a2bd9f09c5ea095884e678fd4458f96 into this branch so that merging it does not break the RichText editor (I don't have the time to find a better fix).

+1

**Comment by michalpipa** [comment]
@dpobel Rebased PR and cherry-picked 658598b81a2bd9f09c5ea095884e678fd4458f96

**Comment by andrerom** [comment]
+1
